File: app/src/main/java/openfoodfacts/github/scrachx/openfood/features/adapters/ProductsRecyclerViewAdapter.java
package openfoodfacts.github.scrachx.openfood.features.adapters;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.ProgressBar;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.squareup.picasso.Callback;
import com.squareup.picasso.Picasso;

import org.apache.commons.lang.StringUtils;

import java.util.List;

import openfoodfacts.github.scrachx.openfood.R;
import openfoodfacts.github.scrachx.openfood.features.productlist.ProductListActivity;
import openfoodfacts.github.scrachx.openfood.models.Product;
import openfoodfacts.github.scrachx.openfood.network.OpenFoodAPIClient;
import openfoodfacts.github.scrachx.openfood.utils.LocaleHelper;
import openfoodfacts.github.scrachx.openfood.utils.Utils;

/**
 * @author herau & itchix
 */
public class ProductsRecyclerViewAdapter extends RecyclerView.Adapter<ProductsRecyclerViewAdapter.ProductsListViewHolder> {
    private static final int VIEW_ITEM = 1;
    private static final int VIEW_LOAD = 0;
    private Context context;
    private final boolean isLowBatteryMode;
    private final List<Product> products;

    public ProductsRecyclerViewAdapter(List<Product> items, boolean isLowBatteryMode) {
        this.products = items;
        this.isLowBatteryMode = isLowBatteryMode;
    }

    @NonNull
    @Override
    public ProductsListViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        context = parent.getContext();

        int layoutResourceId = viewType == VIEW_ITEM ? R.layout.products_list_item : R.layout.progressbar_endless_list;
        View view = LayoutInflater.from(parent.getContext()).inflate(layoutResourceId, parent, false);

        if (viewType == VIEW_ITEM) {
            return new ProductViewHolder(view);
        } else {
            return new ProgressViewHolder(view);
        }
    }

    @Override
    public int getItemViewType(int position) {
        return products.get(position) != null ? VIEW_ITEM : VIEW_LOAD;
    }

    @Override
    public void onBindViewHolder(@NonNull ProductsListViewHolder holder, int position) {
        // - get element from your dataset at this position
        // - replace the contents of the view with that element
        if (!(holder instanceof ProductViewHolder)) {
            return;
        }
        ProductViewHolder productHolder = (ProductViewHolder) holder;
        productHolder.vProductImageProgressbar.setVisibility(View.VISIBLE);
        final String imageSmallUrl = products.get(position).getImageSmallUrl(LocaleHelper.getLanguage(ProductsRecyclerViewAdapter.this.context));
        if (imageSmallUrl == null) {
            productHolder.vProductImageProgressbar.setVisibility(View.GONE);
        }

        // Load Image if isLowBatteryMode is false
        if (!isLowBatteryMode) {
            Utils.picassoBuilder(context)
                .load(imageSmallUrl)
                .placeholder(R.drawable.placeholder_thumb)
                .error(R.drawable.error_image)
                .fit()
                .centerCrop()
                .into(productHolder.vProductImage, new Callback() {
                    @Override
                    public void onSuccess() {
                        productHolder.vProductImageProgressbar.setVisibility(View.GONE);
                    }

                    @Override
                    public void onError(Exception ex) {
                        productHolder.vProductImageProgressbar.setVisibility(View.GONE);
                    }
                });
        } else {
            Picasso.get().load(R.drawable.placeholder_thumb).into(productHolder.vProductImage);
            productHolder.vProductImageProgressbar.setVisibility(View.INVISIBLE);
        }

        Product product = products.get(position);

        productHolder.vProductName.setText(product.getProductName());
        String productNameInLocale = (String) product.getAdditionalProperties().get(OpenFoodAPIClient.getLocaleProductNameField());
        if (StringUtils.isNotBlank(productNameInLocale)) {
            productHolder.vProductName.setText(productNameInLocale);
        }

        String brandsQuantityDetails = ProductListActivity.getProductBrandsQuantityDetails(product);

        final int gradeResource = Utils.getSmallImageGrade(product);
        if (gradeResource != 0) {
            productHolder.vProductGrade.setVisibility(View.VISIBLE);
            productHolder.vProductGrade.setImageResource(gradeResource);
        } else {
            productHolder.vProductGrade.setVisibility(View.INVISIBLE);
        }
        productHolder.vProductDetails.setText(brandsQuantityDetails);
    }

    public Product getProduct(int position) {
        return products.get(position);
    }

    @Override
    public int getItemCount() {
        return products.size();
    }

    /**
     * Provide a reference to the views for each data item
     * Complex data items may need more than one view per item, and
     * you provide access to all the views for a data item in a view holder
     */
    static class ProductViewHolder extends ProductsListViewHolder {
        final TextView vProductDetails;
        final ImageView vProductGrade;
        final ImageView vProductImage;
        final ProgressBar vProductImageProgressbar;
        final TextView vProductName;

        ProductViewHolder(View v) {
            super(v);
            vProductImage = v.findViewById(R.id.imgProduct);
            vProductGrade = v.findViewById(R.id.imgGrade);
            vProductName = v.findViewById(R.id.nameProduct);
            vProductDetails = v.findViewById(R.id.productDetails);
            vProductImageProgressbar = v.findViewById(R.id.searchImgProgressbar);
        }
    }

    static class ProgressViewHolder extends ProductsListViewHolder {
        public ProgressViewHolder(View v) {
            super(v);
        }
    }

    abstract static class ProductsListViewHolder extends RecyclerView.ViewHolder {
        public ProductsListViewHolder(@NonNull View itemView) {
            super(itemView);
        }
    }
}

File: app/src/main/java/openfoodfacts/github/scrachx/openfood/features/adapters/ProductsRecyclerViewAdapter.kt
package openfoodfacts.github.scrachx.openfood.features.adapters

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.ProgressBar
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.squareup.picasso.Callback
import com.squareup.picasso.Picasso
import openfoodfacts.github.scrachx.openfood.AppFlavors
import openfoodfacts.github.scrachx.openfood.R
import openfoodfacts.github.scrachx.openfood.models.Product
import openfoodfacts.github.scrachx.openfood.network.OpenFoodAPIClient.Companion.getLocaleProductNameField
import openfoodfacts.github.scrachx.openfood.utils.LocaleHelper.getLanguage
import openfoodfacts.github.scrachx.openfood.utils.Utils.picassoBuilder
import openfoodfacts.github.scrachx.openfood.utils.getEcoscoreResource
import openfoodfacts.github.scrachx.openfood.utils.getNovaGroupResource
import openfoodfacts.github.scrachx.openfood.utils.getNutriScoreResource
import openfoodfacts.github.scrachx.openfood.utils.getProductBrandsQuantityDetails

/**
 * @author herau & itchix
 */
class ProductsRecyclerViewAdapter(
        val products: MutableList<Product?>,
        private val isLowBatteryMode: Boolean,
        private val context: Context
) : RecyclerView.Adapter<ProductsRecyclerViewAdapter.ProductsListViewHolder>() {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ProductsListViewHolder {
        val layoutResourceId = if (viewType == VIEW_ITEM) R.layout.products_list_item else R.layout.progressbar_endless_list
        val view = LayoutInflater.from(parent.context).inflate(layoutResourceId, parent, false)
        return if (viewType == VIEW_ITEM) ProductsListViewHolder.ProductViewHolder(view) else ProductsListViewHolder.ProgressViewHolder(view)
    }

    override fun getItemViewType(position: Int) = if (products[position] == null) VIEW_LOAD else VIEW_ITEM

    override fun onBindViewHolder(holder: ProductsListViewHolder, position: Int) {
        // - get element from your dataset at this position
        // - replace the contents of the view with that element
        if (holder !is ProductsListViewHolder.ProductViewHolder) return
        holder.imageProgress.visibility = View.VISIBLE
        val imageSmallUrl = products[position]!!.getImageSmallUrl(getLanguage(context))
        if (imageSmallUrl == null) {
            holder.imageProgress.visibility = View.GONE
        }

        // Load Image if isLowBatteryMode is false
        if (!isLowBatteryMode) {
            picassoBuilder(context)
                    .load(imageSmallUrl)
                    .placeholder(R.drawable.placeholder_thumb)
                    .error(R.drawable.error_image)
                    .fit()
                    .centerCrop()
                    .into(holder.productFrontImg, object : Callback {
                        override fun onSuccess() {
                            holder.imageProgress.visibility = View.GONE
                        }

                        override fun onError(ex: Exception) {
                            holder.imageProgress.visibility = View.GONE
                        }
                    })
        } else {
            Picasso.get().load(R.drawable.placeholder_thumb).into(holder.productFrontImg)
            holder.imageProgress.visibility = View.INVISIBLE
        }
        val product = products[position]

        // Set product name
        holder.productName.text = product?.productName ?: context.getString(R.string.productNameNull)
        val productNameInLocale = product?.additionalProperties?.get(getLocaleProductNameField()) as String?
        if (!productNameInLocale.isNullOrBlank()) {
            holder.productName.text = productNameInLocale
        }

        // Set product description
        holder.productDetails.text = product?.getProductBrandsQuantityDetails()

        if (AppFlavors.isFlavors(AppFlavors.OFF)) {
            // Set nutriscore icon
            val nutriRes = product.getNutriScoreResource()
            holder.productNutriscore.setImageResource(nutriRes)
            holder.productNutriscore.visibility = View.VISIBLE

            // Set nova icon
            val novaRes = product.getNovaGroupResource()
            holder.productNova.setImageResource(novaRes)
            holder.productNova.visibility = View.VISIBLE

            // Set ecoscore icon
            val ecoRes = product.getEcoscoreResource()
            holder.productEcoscore.setImageResource(ecoRes)
            holder.productEcoscore.visibility = View.VISIBLE
        }
    }

    fun getProduct(position: Int) = products[position]

    override fun getItemCount() = products.size

    companion object {
        private const val VIEW_ITEM = 1
        private const val VIEW_LOAD = 0
    }

    sealed class ProductsListViewHolder(view: View) : RecyclerView.ViewHolder(view) {
        internal class ProgressViewHolder(view: View) : ProductsListViewHolder(view)

        /**
         * Provide a reference to the views for each data item
         * Complex data items may need more than one view per item, and
         * you provide access to all the views for a data item in a view holder
         */
        internal class ProductViewHolder(view: View) : ProductsListViewHolder(view) {
            val productName: TextView = view.findViewById(R.id.nameProduct)
            val productDetails: TextView = view.findViewById(R.id.productDetails)
            val productNutriscore: ImageView = view.findViewById(R.id.imgNutriscore)
            val productFrontImg: ImageView = view.findViewById(R.id.imgProduct)
            val imageProgress: ProgressBar = view.findViewById(R.id.searchImgProgressbar)
            val productEcoscore: ImageView = view.findViewById(R.id.imgEcoscore)
            val productNova: ImageView = view.findViewById(R.id.imgNova)
        }
    }
}



